circom/tests/subcmps/subcmps4.circom

```diff
@@ -0,0 +1,363 @@
+//CHECK-NEXT:   br label %return3
+//CHECK-EMPTY: 
+//CHECK-NEXT: return3:
+//CHECK-NEXT:   ret void
+//CHECK-NEXT: }
+//
+//CHECK-LABEL: define{{.*}} void @..generated..loop.body.{{[0-9]+\.F\.F}}([0 x i256]* %lvars, [0 x i256]* %signals,
+//CHECK-SAME:  i256* %subfix_[[X1:[0-9]+]], i256* %fix_[[X2:[0-9]+]], i256* %subfix_[[X3:[0-9]+]],
+//CHECK-SAME:  i256* %fix_[[X4:[0-9]+]], [0 x i256]* %sub_[[X3]], i256* %subc_[[X3]]){{.*}} {
+//CHECK-NEXT: ..generated..loop.body.[[$F_ID_3:[0-9]+\.F\.F]]:
+//CHECK-NEXT:   br label %fold_false1
+//CHECK-EMPTY: 
+//CHECK-NEXT: fold_false1:
+//CHECK-NEXT:   %0 = getelementptr i256, i256* %fix_[[X4]], i32 0
+//CHECK-NEXT:   %1 = load i256, i256* %0, align 4
+//CHECK-NEXT:   %2 = getelementptr i256, i256* %subfix_[[X3]], i32 0
+//CHECK-NEXT:   store i256 %1, i256* %2, align 4
+//CHECK-NEXT:   %3 = load i256, i256* %2, align 4
+//CHECK-NEXT:   %constraint = alloca i1, align 1
+//CHECK-NEXT:   call void @__constraint_values(i256 %1, i256 %3, i1* %constraint)
+//CHECK-NEXT:   %4 = load i256, i256* %subc_[[X3]], align 4
+//CHECK-NEXT:   %call.fr_sub = call i256 @fr_sub(i256 %4, i256 1)
+//CHECK-NEXT:   %5 = getelementptr i256, i256* %subc_[[X3]], i32 0
+//CHECK-NEXT:   store i256 %call.fr_sub, i256* %5, align 4
+//CHECK-NEXT:   br label %store2
+//CHECK-EMPTY: 
+//CHECK-NEXT: store2:
+//CHECK-NEXT:   %6 = getelementptr [0 x i256], [0 x i256]* %lvars, i32 0, i32 1
+//CHECK-NEXT:   %7 = load i256, i256* %6, align 4
+//CHECK-NEXT:   %call.fr_add = call i256 @fr_add(i256 %7, i256 1)
+//CHECK-NEXT:   %8 = getelementptr [0 x i256], [0 x i256]* %lvars, i32 0, i32 1
+//CHECK-NEXT:   store i256 %call.fr_add, i256* %8, align 4
+//CHECK-NEXT:   br label %return3
+//CHECK-EMPTY: 
+//CHECK-NEXT: return3:
+//CHECK-NEXT:   ret void
+//CHECK-NEXT: }
+//
+//CHECK-LABEL: define{{.*}} void @..generated..loop.body.{{[0-9]+\.T\.T}}([0 x i256]* %lvars, [0 x i256]* %signals,
+//CHECK-SAME:  i256* %subfix_[[X1:[0-9]+]], i256* %fix_[[X2:[0-9]+]], i256* %subfix_[[X3:[0-9]+]],
+//CHECK-SAME:  i256* %fix_[[X4:[0-9]+]], [0 x i256]* %sub_[[X3]], i256* %subc_[[X3]]){{.*}} {
+//CHECK-NEXT: ..generated..loop.body.[[$F_ID_4:[0-9]+\.T\.T]]:
+//CHECK-NEXT:   br label %fold_true1
+//CHECK-EMPTY: 
+//CHECK-NEXT: fold_true1:
+//CHECK-NEXT:   %0 = getelementptr i256, i256* %fix_[[X2]], i32 0
+//CHECK-NEXT:   %1 = load i256, i256* %0, align 4
+//CHECK-NEXT:   %2 = getelementptr i256, i256* %subfix_[[X1]], i32 0
+//CHECK-NEXT:   store i256 %1, i256* %2, align 4
+//CHECK-NEXT:   %3 = load i256, i256* %2, align 4
+//CHECK-NEXT:   %constraint = alloca i1, align 1
+//CHECK-NEXT:   call void @__constraint_values(i256 %1, i256 %3, i1* %constraint)
+//CHECK-NEXT:   %4 = load i256, i256* %subc_[[X3]], align 4
+//CHECK-NEXT:   %call.fr_sub = call i256 @fr_sub(i256 %4, i256 1)
+//CHECK-NEXT:   %5 = getelementptr i256, i256* %subc_[[X3]], i32 0
+//CHECK-NEXT:   store i256 %call.fr_sub, i256* %5, align 4
+//CHECK-NEXT:   call void @Sum_0_run([0 x i256]* %sub_[[X3]])
```

Comment:
We would have to be careful to only call the run function once even if the output is read more than once or there is more than one output signal.

Reply:
By construction we know that all input signals will be assigned before we can read any outputs so I like Ian's idea in that regards but there are components that sometimes you don't need to read the output (i.e Num2Bits because its constrained within) so we will be missing the run function of that one in some cases